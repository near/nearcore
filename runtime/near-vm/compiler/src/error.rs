use crate::lib::std::string::String;

// Compilation Errors

/// The WebAssembly.CompileError object indicates an error during
/// WebAssembly decoding or validation.
///
/// This is based on the [Wasm Compile Error][compile-error] API.
///
/// [compiler-error]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/CompileError
#[derive(Debug, thiserror::Error)]
pub enum CompileError {
    /// A Wasm translation error occured.
    #[error("WebAssembly translation error: {0}")]
    Wasm(WasmError),

    /// A compilation error occured.
    #[error("Compilation error: {0}")]
    Codegen(String),

    /// The module did not pass validation.
    #[error("Validation error: {0}")]
    Validate(String),

    /// Finite-wasm failed to handle the module.
    #[error("Finite-wasm analysis error: {0}")]
    Analyze(finite_wasm::Error),

    /// The compiler doesn't support a Wasm feature
    #[error("Feature {0} is not yet supported")]
    UnsupportedFeature(String),

    /// The compiler cannot compile for the given target.
    /// This can refer to the OS, the chipset or any other aspect of the target system.
    #[error("The target {0} is not yet supported (see https://docs.wasmer.io/ecosystem/wasmer/wasmer-features)")]
    UnsupportedTarget(String),

    /// Insufficient resources available for execution.
    #[error("Insufficient resources: {0}")]
    Resource(String),

    /// Cannot downcast the engine to a specific type.
    #[error("data offset is out of bounds")]
    InvalidOffset,
}

impl From<WasmError> for CompileError {
    fn from(original: WasmError) -> Self {
        Self::Wasm(original)
    }
}

/// A error in the middleware.
#[derive(Debug, thiserror::Error)]
#[error("Error in middleware {name}: {message}")]
pub struct MiddlewareError {
    /// The name of the middleware where the error was created
    pub name: String,
    /// The error message
    pub message: String,
}

impl MiddlewareError {
    /// Create a new `MiddlewareError`
    pub fn new<A: Into<String>, B: Into<String>>(name: A, message: B) -> Self {
        Self { name: name.into(), message: message.into() }
    }
}

/// A WebAssembly translation error.
///
/// When a WebAssembly function can't be translated, one of these error codes will be returned
/// to describe the failure.
#[derive(Debug, thiserror::Error)]
pub enum WasmError {
    /// The input WebAssembly code is invalid.
    ///
    /// This error code is used by a WebAssembly translator when it encounters invalid WebAssembly
    /// code. This should never happen for validated WebAssembly code.
    #[error("Invalid input WebAssembly code at offset {offset}: {message}")]
    InvalidWebAssembly {
        /// A string describing the validation error.
        message: String,
        /// The bytecode offset where the error occurred.
        offset: usize,
    },

    /// A feature used by the WebAssembly code is not supported by the embedding environment.
    ///
    /// Embedding environments may have their own limitations and feature restrictions.
    #[error("Unsupported feature: {0}")]
    Unsupported(String),

    /// An implementation limit was exceeded.
    #[error("Implementation limit exceeded")]
    ImplLimitExceeded,

    /// An error from the middleware error.
    #[error("{0}")]
    Middleware(MiddlewareError),

    /// A generic error.
    #[error("{0}")]
    Generic(String),
}

impl From<MiddlewareError> for WasmError {
    fn from(original: MiddlewareError) -> Self {
        Self::Middleware(original)
    }
}

/// The error that can happen while parsing a `str`
/// to retrieve a [`CpuFeature`](crate::target::CpuFeature).
#[derive(Debug, thiserror::Error)]
pub enum ParseCpuFeatureError {
    /// The provided string feature doesn't exist
    #[error("CpuFeature {0} not recognized")]
    Missing(String),
}

/// A convenient alias for a `Result` that uses `WasmError` as the error type.
pub type WasmResult<T> = Result<T, WasmError>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn middleware_error_can_be_created() {
        let msg = String::from("Something went wrong");
        let error = MiddlewareError::new("manipulator3000", msg);
        assert_eq!(error.name, "manipulator3000");
        assert_eq!(error.message, "Something went wrong");
    }

    #[test]
    fn middleware_error_be_converted_to_wasm_error() {
        let error = WasmError::from(MiddlewareError::new("manipulator3000", "foo"));
        match error {
            WasmError::Middleware(MiddlewareError { name, message }) => {
                assert_eq!(name, "manipulator3000");
                assert_eq!(message, "foo");
            }
            err => panic!("Unexpected error: {:?}", err),
        }
    }
}
