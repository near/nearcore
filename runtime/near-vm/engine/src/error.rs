//! The WebAssembly possible errors
use crate::trap::RuntimeError;
use near_vm_compiler::CompileError;
use near_vm_types::ExternType;
use std::io;
use thiserror::Error;

/// The Deserialize error can occur when loading a
/// compiled Module from a binary.
#[derive(Error, Debug)]
pub enum DeserializeError {
    /// An IO error
    #[error(transparent)]
    Io(#[from] io::Error),
    /// A generic deserialization error
    #[error("{0}")]
    Generic(String),
    /// Incompatible serialized binary
    #[error("incompatible binary: {0}")]
    Incompatible(String),
    /// The provided binary is corrupted
    #[error("corrupted binary: {0}")]
    CorruptedBinary(String),
    /// The binary was valid, but we got an error when
    /// trying to allocate the required resources.
    #[error(transparent)]
    Compiler(CompileError),
}

/// An ImportError.
///
/// Note: this error is not standard to WebAssembly, but it's
/// useful to determine the import issue on the API side.
#[derive(Error, Debug)]
pub enum ImportError {
    /// Incompatible Import Type.
    /// This error occurs when the import types mismatch.
    #[error("incompatible import type. Expected {0:?} but received {1:?}")]
    IncompatibleType(ExternType, ExternType),

    /// Unknown Import.
    /// This error occurs when an import was expected but not provided.
    #[error("unknown import. Expected {0:?}")]
    UnknownImport(ExternType),
}

/// The WebAssembly.LinkError object indicates an error during
/// module instantiation (besides traps from the start function).
///
/// This is based on the [link error][link-error] API.
///
/// [link-error]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/LinkError
#[derive(Error, Debug)]
#[error("Link error: {0}")]
pub enum LinkError {
    /// An error occurred when checking the import types.
    #[error("Error while importing {0:?}.{1:?}: {2}")]
    Import(String, String, ImportError),

    /// A trap ocurred during linking.
    #[error("RuntimeError occurred during linking: {0}")]
    Trap(#[source] RuntimeError),

    /// Insufficient resources available for linking.
    #[error("Insufficient resources: {0}")]
    Resource(String),
}

/// An error while instantiating a module.
///
/// This is not a common WebAssembly error, however
/// we need to differentiate from a `LinkError` (an error
/// that happens while linking, on instantiation) and a
/// Trap that occurs when calling the WebAssembly module
/// start function.
#[derive(Error, Debug)]
pub enum InstantiationError {
    /// A linking ocurred during instantiation.
    #[error(transparent)]
    Link(LinkError),

    /// The module was compiled with a CPU feature that is not available on
    /// the current host.
    #[error("module compiled with CPU feature that is missing from host")]
    CpuFeature(String),

    /// A runtime error occured while invoking the start function
    #[error(transparent)]
    Start(RuntimeError),
}
