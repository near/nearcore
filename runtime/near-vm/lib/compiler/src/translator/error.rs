use crate::{CompileError, WasmError};
use wasmparser::BinaryReaderError;

/// Return an `Err(WasmError::Unsupported(msg))` where `msg` the string built by calling `format!`
/// on the arguments to this macro.
#[macro_export]
macro_rules! wasm_unsupported {
    ($($arg:tt)*) => { $crate::WasmError::Unsupported(format!($($arg)*)) }
}

impl From<BinaryReaderError> for WasmError {
    fn from(original: BinaryReaderError) -> Self {
        Self::InvalidWebAssembly { message: original.message().into(), offset: original.offset() }
    }
}

impl From<BinaryReaderError> for CompileError {
    fn from(original: BinaryReaderError) -> Self {
        // `From` does not seem to be transitive by default, so we convert
        // BinaryReaderError -> WasmError -> CompileError
        Self::from(WasmError::from(original))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use wasmparser::BinaryReader;

    #[test]
    fn can_convert_binary_reader_error_to_wasm_error() {
        let mut reader = BinaryReader::new(b"\0\0\0\0");
        let binary_reader_error = reader.read_bytes(10).unwrap_err();
        match WasmError::from(binary_reader_error) {
            WasmError::InvalidWebAssembly { message, offset } => {
                assert_eq!(message, "unexpected end-of-file");
                assert_eq!(offset, 0);
            }
            err => panic!("Unexpected error: {:?}", err),
        }
    }

    #[test]
    fn can_convert_binary_reader_error_to_compile_error() {
        let mut reader = BinaryReader::new(b"\0\0\0\0");
        let binary_reader_error = reader.read_bytes(10).unwrap_err();
        match CompileError::from(binary_reader_error) {
            CompileError::Wasm(WasmError::InvalidWebAssembly { message, offset }) => {
                assert_eq!(message, "unexpected end-of-file");
                assert_eq!(offset, 0);
            }
            err => panic!("Unexpected error: {:?}", err),
        }
    }
}
