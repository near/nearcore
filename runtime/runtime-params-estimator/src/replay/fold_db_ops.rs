use super::cache_stats::CacheStats;
use super::Visitor;
use std::collections::{BTreeMap, HashMap};
use std::io::Write;

const EMPTY_STATE_ERR: &str = "states must never be empty";

/// A visitor that keeps track of DB operations and aggregates it by specific labels.
pub(super) struct FoldDbOps {
    /// Labels at which to break aggregation, and the fields to print on each.
    ///
    /// If field is not available, it will be silently ignored.
    fold_anchors: HashMap<String, Vec<String>>,
    /// Print the DB operations that are at indentation 0, not associated with a span.
    ///
    /// This is a summary of the total. Instead it shows what operations have
    /// not been covered by already reported output. To get a summary of the
    /// total, run without folding.
    print_top_level: bool,
    /// Only show data for a specific smart contract, specified by account id.
    /// Currently only applies within receipts.
    account_filter: Option<String>,
    /// Only evaluate lines with at least this depth of indentation.
    ///
    /// Used to apply filters and skip some parts of the trace.
    /// Default is 0, which collects everything.
    min_indent: usize,
    /// Reset the filter when indentation goes back to this value.
    ///
    /// Used in combination with `min_indent` and `account_filter`.
    /// Resetting means that `min_indent` is set to usize::MAX and nothing will
    /// be evaluated until the `account_filter` is triggered again and sets it
    /// to a smaller value again.
    filter_reset_indent: Option<usize>,
    /// Optionally collect and print detailed statistics for cache hits and misses.
    track_caches: bool,
    /// Keeps track of current block.
    block_hash: Option<String>,
    /// Stack of states, each starting at a specific indent.
    states: Vec<State>,
}

/// Stores statistics for a range of the trace, starting at a fixed indent and
/// collecting data for everything indented further, until another state object
/// is pushed on top.
#[derive(Default)]
struct State {
    /// The indent at which this state started
    indent: usize,
    /// Keeps track of operations per DB column.
    ops_cols: BTreeMap<String, BTreeMap<String, usize>>,
    /// Optionally collect and print detailed statistics for cache hits and misses.
    cache_stats: Option<CacheStats>,
}

impl FoldDbOps {
    pub(super) fn new() -> Self {
        Self {
            fold_anchors: HashMap::new(),
            print_top_level: true,
            account_filter: None,
            track_caches: false,
            states: vec![State::default()],
            block_hash: None,
            min_indent: 0,
            filter_reset_indent: None,
        }
    }

    /// Pre-set that folds on chunks.
    pub(super) fn chunks(self) -> Self {
        self.fold("apply_transactions", &["receiver", "shard_id"])
    }

    /// Pre-set that folds on receipts.
    pub(super) fn receipts(self) -> Self {
        self.fold("process_receipt", &["receiver", "receipt_id"])
            .fold("process_transaction", &["tx_hash"])
            .print_top_level(false)
    }

    pub(super) fn fold(mut self, anchor: impl Into<String>, printed_fields: &[&str]) -> Self {
        let fields = printed_fields.into_iter().map(|s| (*s).into()).collect();
        self.fold_anchors.insert(anchor.into(), fields);
        self
    }

    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
        self.print_top_level = yes;
        self
    }

    pub(super) fn with_cache_stats(mut self) -> Self {
        self.track_caches = true;
        self.state().cache_stats = Some(CacheStats::default());
        self
    }

    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
        if account.is_some() {
            // evaluate nothing if there is a filter, until the filter matches the first time
            self.min_indent = usize::MAX;
        }
        self.account_filter = account;
        self
    }

    fn state(&mut self) -> &mut State {
        self.states.last_mut().expect(EMPTY_STATE_ERR)
    }

    fn push_state(&mut self, indent: usize) {
        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
        self.states.push(new_state);
    }

    fn pop_state(&mut self) -> State {
        let state = self.states.pop().expect(EMPTY_STATE_ERR);
        if self.states.is_empty() {
            self.push_state(0);
        }
        state
    }

    fn skip_eval(&mut self, trace_indent: usize) -> bool {
        trace_indent < self.min_indent
    }

    /// Check if indentation has gone back enough to pop current state or reset filter.
    ///
    /// Call this before `skip()` to ensure it uses the correct `min_indent`.
    fn update_state(&mut self, out: &mut dyn Write, indent: usize) -> anyhow::Result<()> {
        if self.states.len() > 1 && self.state().indent >= indent {
            self.pop_state().print(out)?;
        }
        if let Some(reset_indent) = self.filter_reset_indent {
            if indent <= reset_indent {
                self.filter_reset_indent = None;
                self.min_indent = usize::MAX;
            }
        }
        Ok(())
    }
}

impl Visitor for FoldDbOps {
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        self.update_state(out, indent)?;

        // Block hash in traces is visibly by looking at DB lookups of
        // `BlockInfo` column. Keep track of it so that each time something is
        // printed, the block hash can be included if desired.
        match op {
            "GET" => {
                if col == "BlockInfo" {
                    self.block_hash = Some(bs58::encode(key).into_string());
                }
            }
            _ => {
                // nop
            }
        }

        if self.skip_eval(indent) {
            return Ok(());
        }

        // Count DB operation for the corresponding column.
        *self
            .state()
            .ops_cols
            .entry(op.to_owned())
            .or_default()
            .entry(col.to_owned())
            .or_default() += 1;

        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_db_op(op, size);
        }

        Ok(())
    }

    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.update_state(out, indent)?;
        if self.skip_eval(indent) {
            return Ok(());
        }
        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_storage_op(op, dict)?;
        }
        Ok(())
    }

    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.update_state(out, indent)?;
        if let (Some(receiver), Some(filtered_account)) =
            (dict.get("receiver"), &self.account_filter)
        {
            if receiver == filtered_account && self.filter_reset_indent.is_none() {
                self.min_indent = indent;
                self.filter_reset_indent = Some(indent);
            }
        }
        if self.skip_eval(indent) {
            return Ok(());
        }
        if self.fold_anchors.contains_key(label) {
            // Section to fold on starts. Push a new state on the stack and
            // print the header for the new section.
            self.push_state(indent);
            write!(out, "{:indent$}{label}", "")?;
            // Unnecessary perf optimization: Second lookup in fold anchors
            // could be avoided by reading the key directly but then we keep
            // a mutable reference to self and cannot naively call
            // self.push_state above.
            // Better to lookup twice and keep code simple.
            for key in self.fold_anchors.get(label).expect("just checked contains key").iter() {
                if let Some(value) = dict.get(key.as_str()) {
                    write!(out, " {key}={value}")?;
                }
            }
            if let Some(block) = &self.block_hash {
                write!(out, " block={block}")?;
            }
            writeln!(out)?;
        }

        if let Some(cache_stats) = &mut self.state().cache_stats {
            cache_stats.eval_generic_label(dict);
        }
        Ok(())
    }

    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        if self.print_top_level {
            writeln!(out, "top-level:")?;
            self.pop_state().print(out)?;
        }
        Ok(())
    }
}

impl State {
    fn print(self, out: &mut dyn Write) -> anyhow::Result<()> {
        let indent = self.indent + 2;
        for (op, map) in self.ops_cols.into_iter() {
            if !map.is_empty() {
                write!(out, "{:indent$}{op}   ", "")?;
            }
            for (col, num) in map.into_iter() {
                write!(out, "{num:8>} {col}  ")?;
            }
            writeln!(out)?;
        }

        if let Some(stats) = self.cache_stats {
            stats.print(out, indent)?;
        }
        writeln!(out)?;
        Ok(())
    }
}
