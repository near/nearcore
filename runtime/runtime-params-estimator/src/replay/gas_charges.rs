use super::Visitor;
use std::collections::BTreeMap;
use std::io::Write;

/// Visitor that tracks which DB operations are charged with gas and which are
/// for free.
#[derive(Default)]
pub(super) struct ChargedVsFree {
    in_fn_call: bool,
    storage_op_indent: usize,

    free_gets: u64,
    free_gets_size: u64,
    charged_gets: u64,
    charged_gets_size: u64,
}

impl Visitor for ChargedVsFree {
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        _key: &[u8],
        _col: &str,
    ) -> anyhow::Result<()> {
        self.eval_label(out, indent, op, &BTreeMap::new())?;
        if op != "GET" {
            return Ok(());
        }
        if self.in_fn_call {
            self.charged_gets += 1;
            self.charged_gets_size += size.unwrap_or(0);
        } else {
            self.free_gets += 1;
            self.free_gets_size += size.unwrap_or(0);
        }
        Ok(())
    }

    fn eval_storage_op(
        &mut self,
        _out: &mut dyn Write,
        indent: usize,
        _op: &str,
        _dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.storage_op_indent = indent + 2;
        self.in_fn_call = true;
        Ok(())
    }

    fn eval_label(
        &mut self,
        _out: &mut dyn Write,
        indent: usize,
        _label: &str,
        _dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        if indent < self.storage_op_indent {
            self.in_fn_call = false;
        }
        Ok(())
    }

    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        writeln!(
            out,
            "{:>8} free gets with total size of    {:>8}",
            self.free_gets, self.free_gets_size
        )?;
        writeln!(
            out,
            "{:>8} charged gets with total size of {:>8}",
            self.charged_gets, self.charged_gets_size
        )?;
        writeln!(
            out,
            "{:>7.2}% of gets uncharged",
            100.0 * self.free_gets as f64 / (self.charged_gets + self.free_gets) as f64
        )?;
        writeln!(
            out,
            "{:>7.2}% of gets total size uncharged",
            100.0 * self.free_gets_size as f64
                / (self.charged_gets_size + self.free_gets_size) as f64
        )?;

        self.free_gets = 0;
        self.free_gets_size = 0;
        self.charged_gets = 0;
        self.charged_gets_size = 0;

        Ok(())
    }
}
