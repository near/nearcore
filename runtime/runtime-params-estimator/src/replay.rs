use anyhow::Context;
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, Write};
use std::path::PathBuf;
use std::str::SplitWhitespace;
use tracing::log::error;

use self::fold_db_ops::FoldDbOps;
use self::gas_charges::ChargedVsFree;

mod cache_stats;
mod fold_db_ops;
mod gas_charges;

#[derive(clap::Parser)]
pub(crate) struct ReplayCmd {
    trace: PathBuf,
    #[clap(subcommand)]
    mode: ReplayMode,
    /// Only show data for a specific smart contract, specified by account id.
    #[clap(long)]
    account: Option<String>,
}

#[derive(Clone, Copy, clap::Subcommand, Debug)]
pub(crate) enum ReplayMode {
    /// Print DB accesses and cache statistics for the entire trace.
    CacheStats,
    /// Print DB accesses per receipt.
    ReceiptDbStats,
    /// Print DB accesses and cache statistics per receipt.
    ReceiptCacheStats,
    /// Print DB accesses per chunk.
    ChunkDbStats,
    /// Print DB accesses and cache statistics per chunk.
    ChunkCacheStats,
    /// Go over DB operations and print how much of it is paid for with gas.
    GasCharges,
}

impl ReplayCmd {
    pub(crate) fn run(&self, out: &mut dyn Write) -> anyhow::Result<()> {
        let file = File::open(&self.trace)?;
        self.run_on_input(io::BufReader::new(file), out)
    }

    fn run_on_input(&self, input: impl io::BufRead, out: &mut dyn Write) -> anyhow::Result<()> {
        let mut visitor = self.build_visitor();
        for line in input.lines() {
            let line = line?;
            if let Err(e) = visitor.eval_line(out, &line) {
                error!("ERROR: {e} for input line: {line}");
            }
        }
        visitor.flush(out)?;
        Ok(())
    }

    fn build_visitor(&self) -> Box<dyn Visitor> {
        match &self.mode {
            ReplayMode::CacheStats => {
                Box::new(FoldDbOps::new().with_cache_stats().account_filter(self.account.clone()))
            }
            ReplayMode::ChunkDbStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks())
            }
            ReplayMode::ChunkCacheStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks().with_cache_stats())
            }
            ReplayMode::ReceiptDbStats => {
                Box::new(FoldDbOps::new().receipts().account_filter(self.account.clone()))
            }
            ReplayMode::ReceiptCacheStats => Box::new(
                FoldDbOps::new().receipts().with_cache_stats().account_filter(self.account.clone()),
            ),
            ReplayMode::GasCharges => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with gas charges");
                }
                Box::new(ChargedVsFree::default())
            }
        }
    }
}

/// Interface for processing an IO trace line-by-line.
///
/// The visitor default methods take over the basic parsing of the IO trace.
/// Specific implementations can extract the data of interest and process it
/// further as necessary, without duplicating the parsing code.
trait Visitor {
    /// The root entry point of the visitors.
    ///
    /// This function takes a raw input line as input without any preprocessing.
    /// A visitor may choose to overwrite this function for full control but the
    /// intention is that the default implementation takes over the basic
    /// parsing and visitor implementations define their behaviour using the
    /// other trait methods.
    fn eval_line(&mut self, out: &mut dyn Write, line: &str) -> anyhow::Result<()> {
        if let Some(indent) = line.chars().position(|c| !c.is_whitespace()) {
            let mut tokens = line.split_whitespace();
            if let Some(keyword) = tokens.next() {
                match keyword {
                    "GET" | "SET" | "UPDATE_RC" => {
                        let col = tokens.next().context("missing column field in DB operation")?;
                        let mut key_str = tokens.next().context("missing key in DB operation")?;
                        if key_str.starts_with('"') {
                            key_str = &key_str[1..key_str.len() - 1];
                        }
                        let key = near_fmt::Bytes::from_str(key_str).map_err(|e| {
                            anyhow::anyhow!("failed to decode key {key_str} because {e}")
                        })?;
                        let dict = extract_key_values(tokens)?;
                        let size: Option<u64> = dict.get("size").map(|s| s.parse()).transpose()?;
                        self.eval_db_op(out, indent, keyword, size, &key, col)?;
                    }
                    "storage_read" | "storage_write" | "storage_remove" | "storage_has_key" => {
                        let op = tokens.next();
                        if op.is_none() {
                            return Ok(());
                        }

                        let dict = extract_key_values(tokens)?;
                        self.eval_storage_op(out, indent, keyword, &dict)?;
                    }
                    other_label => {
                        let dict = extract_key_values(tokens)?;
                        self.eval_label(out, indent, other_label, &dict)?;
                    }
                }
            }
        }
        Ok(())
    }

    /// Gets called for every trie storage operation.
    ///
    /// A storage operation is a layer above DB operations, one storage
    /// operation can cause many DB operations, or it could be resolved from
    /// cache without any DB operations.
    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, op, dict);
        Ok(())
    }

    /// Gets called for every DB operation.
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        if col == "State" {
            self.eval_state_db_op(out, indent, op, size, key)
        } else {
            Ok(())
        }
    }

    /// Gets called for every DB operation on the state column.
    fn eval_state_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
    ) -> anyhow::Result<()> {
        let (_, _, _, _, _) = (out, indent, op, size, key);
        Ok(())
    }

    /// Opening spans that are not storage or DB operations.
    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, label, dict);
        Ok(())
    }

    /// Print accumulated internal state.
    ///
    /// This function is called at the end of a replayed trace. Visitor
    /// implementation usually choose to call this more often. For example, once
    /// for each receipt.
    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        let _ = out;
        Ok(())
    }
}

fn extract_key_values<'a>(
    mut tokens: SplitWhitespace<'a>,
) -> anyhow::Result<BTreeMap<&'a str, &'a str>> {
    let mut dict = BTreeMap::new();
    while let Some(key_val) = tokens.next() {
        let (key, value) =
            key_val.split_once('=').context("key-value pair delimited by `=` expected")?;
        dict.insert(key, value);
    }
    Ok(dict)
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::{ReplayCmd, ReplayMode};

    /// These inputs are real mainnet traffic for the given block heights.
    /// Each trace contains two chunks in one shard.
    /// In combination, they cover an interesting variety of content.
    ///  Shard 0: Many receipts, but no function calls.
    ///  Shard 1: Empty chunks without txs / receipts.
    ///  Shard 2: Some receipts and txs, but no function calls.
    ///  Shard 3: Many function calls with lots of storage accesses.
    const INPUT_TRACES: &[&str] = &[
        "75220100-75220101.s0.io_trace",
        "75220100-75220101.s1.io_trace",
        "75220100-75220101.s2.io_trace",
        "75220100-75220101.s3.io_trace",
    ];

    /// A synthetic trace that should be a bit easier to read and verify.
    const SYNTHETIC_TRACE: &str = r#"
GET BlockHeader "`fAkeHeAd3R`" size=6000
GET BlockInfo "`FAk31nf0`" size=1
apply_transactions shard_id=0
    process_state_update 
        apply num_transactions=1 shard_cache_hit=10 shard_cache_miss=1
            process_transaction tx_hash=txHash0 shard_cache_miss=1 shard_cache_hit=20
                GET State "'stateKey0'" size=300
            process_receipt receipt_id=id0 predecessor=system receiver=alice.near id=id0 shard_cache_miss=2 shard_cache_hit=6
                GET State "'stateKey1'" size=10
                GET State "'stateKey2'" size=20
            process_receipt receipt_id=id1 predecessor=system receiver=bob.near id=id1 shard_cache_hit=18 shard_cache_miss=3 shard_cache_too_large=1
                GET State "'stateKey3'" size=100
                GET State "'stateKey4'" size=200
                GET State "'stateKey5'" size=9000
            process_receipt receipt_id=id2 predecessor=system receiver=alice.near id=id2 shard_cache_miss=1 shard_cache_hit=6
                GET State "'stateKey6'" size=30
                attached_deposit 
                input 
                register_len 
                read_register 
                storage_read READ key=StorageKey0 size=1000 tn_db_reads=20 tn_mem_reads=0 shard_cache_hit=19 shard_cache_miss=1
                    GET State "'stateKey7'" size=5
            process_receipt receipt_id=id3 predecessor=utiha.near receiver=alice.near id=id3 shard_cache_miss=0 shard_cache_hit=15
            GET State "'stateKey8'" size=300
        GET State "'stateKey9'" size=400
GET State "'stateKey10'" size=500
"#;

    #[test]
    fn test_cache_stats() {
        check_replay_mode(ReplayMode::CacheStats);
    }

    #[test]
    fn test_chunk_cache_stats() {
        check_replay_mode(ReplayMode::ChunkCacheStats);
    }

    #[test]
    fn test_chunk_db_stats() {
        check_replay_mode(ReplayMode::ChunkDbStats);
    }

    #[test]
    fn test_gas_charges() {
        check_replay_mode(ReplayMode::GasCharges);
    }

    #[test]
    fn test_receipt_cache_stats() {
        check_replay_mode(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_receipt_db_stats() {
        check_replay_mode(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_replay_mode(mode: ReplayMode) {
        for trace_name in INPUT_TRACES {
            let dir = env!("CARGO_MANIFEST_DIR");
            let trace_path = std::path::Path::new(dir).join("res").join(trace_name);
            let cmd = ReplayCmd { trace: trace_path, mode, account: None };
            let mut buffer = Vec::new();
            cmd.run(&mut buffer).unwrap_or_else(|e| {
                panic!("command should not fail for input {trace_name}, failure was {e}")
            });
            let output = String::from_utf8(buffer).unwrap_or_else(|e| {
                panic!("invalid output for input {trace_name}, failure was {e}")
            });
            insta::assert_snapshot!(format!("{mode:?}-{trace_name}"), output);
        }
    }

    #[test]
    fn test_account_filter_cache_stats() {
        check_account_filter(ReplayMode::CacheStats);
    }

    #[test]
    fn test_account_filter_receipt_cache_stats() {
        check_account_filter(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_account_filter_receipt_db_stats() {
        check_account_filter(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_account_filter(mode: ReplayMode) {
        let account = Some("alice.near".to_owned());
        // trace path not used, will be read from in-memory input instead
        let trace = PathBuf::new();
        let cmd = ReplayCmd { trace, mode, account };
        let mut buffer = Vec::new();
        cmd.run_on_input(SYNTHETIC_TRACE.as_bytes(), &mut buffer).expect("failed replaying");
        let output =
            String::from_utf8(buffer).unwrap_or_else(|e| panic!("invalid output, failure was {e}"));
        insta::assert_snapshot!(format!("account_filter_{mode:?}"), output);
    }
}
