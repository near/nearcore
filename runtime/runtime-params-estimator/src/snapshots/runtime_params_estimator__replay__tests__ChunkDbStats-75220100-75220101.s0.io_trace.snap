---
source: runtime/runtime-params-estimator/src/replay.rs
expression: output
---
apply_transactions shard_id=0 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
  GET   287 State  

apply_transactions shard_id=0 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
  GET   345 State  

top-level:
  GET   4 Block  4 BlockHeader  2 BlockHeight  4 BlockInfo  2 BlockMisc  2 ChunkExtra  2 Chunks  4 EpochInfo  2 IncomingReceipts  4 State  


